# Project

# 1. About Dataset 

## 1-1) 데이터 선정사유 
  

최근 기후 변화로 와인의 생산량이 급감하고 있는 문제가 발생되고 있다.  
이러한 현상으로 와인 풍귀현상 및 사재기 현상도 함께 발생되어 사회문제로 부각되고 있다.  

이에, 미국, 프랑스, 이테리, 대한민국 정부에서는 와인 판매를 국가에서 통제하고자 한다.  
정부 담당자는 판매되는 와인의 가격으로 결정하기 위해 기존에 판매된 와인가격에 대한 데이터 분석을 요청하였다.  
이와 더불어 공정성에 매우 민감한 문제이므로 충분한 데이터(10만건) 이상 분석을 요청하였습니다. 담당자는 캐글에서 10만건 이상되는 데이터가 필요하여 해당 데이터를 선정하였습니다. 


DS.ltd 와인 가격분석을 진행하고자 합니다.  


---

## 1-2) target 선정사유 

  
우리는 기존에 판매된 와인가격에 대한 분석을 진행해야 하므로 price를 target으로 선정토록 하겠습니다.  
  
또한, price를 기준으로 어떤 와인이 좋다 나쁘다의 분류를 하는 것이 아니며   
적절한 가격은 얼마정도인지 가격 예측을 목표로 하기 때문에 회귀분석으로 접근토록 하겠습니다. 

---

## 1-3) 데이터 로드 및 확인  

In [1]:
import pandas as pd
import numpy as np 
import re

In [2]:
df = pd.read_csv(r"C:\Users\user\Documents\pandas_study\bootcamp_DS\project\section2_porject_210222\winemag-data-130k-v2.csv", index_col=0)

In [3]:
df.head(50)

country                                        description  \
0       Italy  Aromas include tropical fruit, broom, brimston...   
1    Portugal  This is ripe and fruity, a wine that is smooth...   
2          US  Tart and snappy, the flavors of lime flesh and...   
3          US  Pineapple rind, lemon pith and orange blossom ...   
4          US  Much like the regular bottling from 2012, this...   
5       Spain  Blackberry and raspberry aromas show a typical...   
6       Italy  Here's a bright, informal red that opens with ...   
7      France  This dry and restrained wine offers spice in p...   
8     Germany  Savory dried thyme notes accent sunnier flavor...   
9      France  This has great depth of flavor with its fresh ...   
10         US  Soft, supple plum envelopes an oaky structure ...   
11     France  This is a dry wine, very spicy, with a tight, ...   
12         US  Slightly reduced, this wine offers a chalky, t...   
13      Italy  This is dominated by oak and oak-driven aromas...   
14         US  Building on 150 years and six generations of w...   
15    Germany  Zesty orange peels and apple notes abound in t...   
16  Argentina  Baked plum, molasses, balsamic vinegar and che...   
17  Argentina  Raw black-cherry aromas are direct and simple ...   
18      Spain  Desiccated blackberry, leather, charred wood a...   
19         US  Red fruit aromas pervade on the nose, with cig...   
20         US  Ripe aromas of dark berries mingle with ample ...   
21         US  A sleek mix of tart berry, stem and herb, alon...   
22      Italy  Delicate aromas recall white flower and citrus...   
23         US  This wine from the Geneseo district offers aro...   
24      Italy  Aromas of prune, blackcurrant, toast and oak c...   
25         US  Oak and earth intermingle around robust aromas...   
26      Italy  Pretty aromas of yellow flower and stone fruit...   
27      Italy  Aromas recall ripe dark berry, toast and a whi...   
28      Italy  Aromas suggest mature berry, scorched earth, a...   
29         US  Clarksburg is becoming a haven for Chenin Blan...   
30     France  Red cherry fruit comes laced with light tannin...   
31      Italy  Merlot and Nero d'Avola form the base for this...   
32      Italy  Part of the extended Calanìca series, this Gri...   
33         US  Rustic and dry, this has flavors of berries, c...   
34         US  This shows a tart, green gooseberry flavor tha...   
35         US  As with many of the Erath 2010 vineyard design...   
36      Chile  White flower, lychee and apple aromas carry th...   
37      Italy  This concentrated Cabernet offers aromas of cu...   
38      Italy  Inky in color, this wine has plump aromas of r...   
39      Italy  Part of the natural wine movement, this wine i...   
40      Italy  Catarratto is one of Sicily's most widely farm...   
41         US  A stiff, tannic wine, this slowly opens and br...   
42     France  This is a festive wine, with soft, ripe fruit ...   
43         US  The clean, brisk mouthfeel gives this slightly...   
44      Chile  A berry aroma comes with cola and herb notes. ...   
45         US  Right out of the starting blocks this is an oa...   
46      Italy  Spicy, fresh and clean, this would pair with f...   
47         US  This is a sweet wine with flavors of white sug...   
48         US  This bottling resembles the New Zealand paradi...   
49     France  Soft and fruity, this is a generous, ripe wine...   

                                          designation  points  price  \
0                                        Vulkà Bianco      87    NaN   
1                                            Avidagos      87   15.0   
2                                                 NaN      87   14.0   
3                                Reserve Late Harvest      87   13.0   
4                  Vintner's Reserve Wild Child Block      87   65.0   
5                                        Ars In Vitro      87   15.0   
6                                

In [4]:
# 데이터량 확인
df.shape

(129971, 13)

In [5]:
# 데이터 결측치 확인 
df.isnull().sum()

country                     63
description                  0
designation              37465
points                       0
price                     8996
province                    63
region_1                 21247
region_2                 79460
taster_name              26244
taster_twitter_handle    31213
title                        0
variety                      1
winery                       0
dtype: int64

In [6]:
# df1.info로 확인 결과 대부분 object 데이터 타입임을 확인하였다. 
# check the data info -> This data is almost object types 
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 129971 entries, 0 to 129970
Data columns (total 13 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   country                129908 non-null  object 
 1   description            129971 non-null  object 
 2   designation            92506 non-null   object 
 3   points                 129971 non-null  int64  
 4   price                  120975 non-null  float64
 5   province               129908 non-null  object 
 6   region_1               108724 non-null  object 
 7   region_2               50511 non-null   object 
 8   taster_name            103727 non-null  object 
 9   taster_twitter_handle  98758 non-null   object 
 10  title                  129971 non-null  object 
 11  variety                129970 non-null  object 
 12  winery                 129971 non-null  object 
dtypes: float64(1), int64(1), object(11)
memory usage: 13.9+ MB


In [7]:
# object data의 unique를 살펴보았다. 
df.describe(exclude='number').T.sort_values(by='unique')

count  unique  \
taster_twitter_handle   98758      15   
region_2                50511      17   
taster_name            103727      19   
country                129908      43   
province               129908     425   
variety                129970     707   
region_1               108724    1229   
winery                 129971   16757   
designation             92506   37979   
title                  129971  118840   
description            129971  119955   

                                                                     top  \
taster_twitter_handle                                         @vossroger   
region_2                                                   Central Coast   
taster_name                                                   Roger Voss   
country                                                               US   
province                                                      California   
variety                                                       Pinot Noir   
region_1                                                     Napa Valley   
winery                                                Wines & Winemakers   
designation                                                      Reserve   
title                  Gloria Ferrer NV Sonoma Brut Sparkling (Sonoma...   
description            Stalky aromas suggest hay and green herbs, wit...   

                        freq  
taster_twitter_handle  25514  
region_2               11065  
taster_name            25514  
country                54504  
province               36247  
variety                13272  
region_1                4480  
winery                   222  
designation             2009  
title                     11  
description                3

## 1-4) feature 설명 
  
'country' - 어느나라에서 생산되었는지  
'description' - 와인 시음 후 와인에 대한 평가 및 설명   
'designation' - 양조장이 있는 포도원   
'points' - 시음 후 와인 점수   
'price' - 와인의 가격 ( "our target" )  
'province' - 와인 제작 시, 사용되는 포도의 출처   
'region_1' - 포도가 자라는 지역 
'region_2' - 포도가 자라는 지역을 상세하게 기재한 feature
'taster_name' - 시음한 사람     
'taster_twitter_handle' 시음한 사람의 트위터   
'title' - 와인명  
'variety' - 와인을 만들때 혼합된 포도  
'winery' - 양조장명  

### **☆ 지금까지 insight ☆**  
 1. designation과 winery를 동일하게 양조장에 대한 정보를 나타내고 있다.  
     -> 확인 후 합칠 수 있는 것들은 혼합할 예정   
       
 2. region_1과 region_2도 동일한 정보를 나타내며, title에 region정보가 있는 것을 확인할 수 있었다.  
    -> 결측치는 이로 대체 예정 
      
 3. taster_name와 taster_twitter_handle도 taster에 대한 동일한 정보를 제공해주고 있다.  
    -> 확인 후 합칠 수 있는 것들은 혼합할 예정 
      
 4. title, variety, winery 등은 결측치가 거의 없다. 이에, 해당 데이터를 활용하여 결측치를 채울 수 있는 것들은 채울 예정
   
 5. description을 split하여 aroma, blend, sweet 등의 단어를 찾아서 타겟과의 연관관계를 확인해보고자 한다    
   
 6. 또한, 너무 unique가 높기 때문에 각 feature별로 count하여 count된 만큼 페센트로 가중치를 주고자 한다.  


### ☆ 나라별 용어 따른 숙성 정도 차이 ☆

7. 스페인, 이탈리아 ( 법적 필수 준수사항 )


 |등급|내용|
 |----|-----|
 |joven|전혀 숙성을 시키지 않고 바로 출시된 와인|
 |crianza|최소 24개월이상 오크 및 병에 담아 숙성 후 출시된 와인|
 |reserva|최소 36개월이상 오크 및 병에 담아 숙성 후 출시된 와인|
 |gran reserva|최소 60개월이상 오크 및 병에 담아 숙성 후 출시된 와인|

8. 그외나라(미국, 칠레, 호주, 뉴질랜드 등)   
 -. 별다른 규정 없이 바로 taster에게 좋은 평가를 받으면 reserva라는 표기 가능   

9. 기타(혼용되는 용어)  
 -. reserva는 나라에 따라서 riserva, reserve 등으로 혼용되어 사용된다고 한다.  
 -. ["À", "A"],["Á", "A"],["Â", "A"],["Ã", "A"],["Ä", "A"],["Ä", "A"],["Å", "A"]  
 -. ,["à", "a"],["á", "a"],["â", "a"],["ã", "a"],["ä", "a"],["ä", "a"],["å", "a"]  
 -. ,["È", "E"],["É", "E"],["Ê", "E"],["Ë", "E"],["Ë", "E"]  
 -. ,["è", "e"],["é", "e"],["ê", "e"],["ë", "e"],["ë", "e"]  
 -. ,["Æ", "AE"],["æ", "ae"],["Ç", "C"],["ç", "c"]  
 -. ,["Ì", "I"],["Í", "I"],["Î", "I"],["Ï", "I"],["Ï", "I"]  
 -. ,["ì", "i"],["í", "i"],["î", "i"],["ï", "i"],["ï", "i"]  
 -. ,["Ð", "D"],["ð", "d"],["Ñ", "N"],["ñ", "n"]  
 -. ,["Ò", "O"],["Ó", "O"],["Ô", "O"],["Õ", "O"],["Ö", "O"],["Ö", "O"],["Œ", "OE"]   
 -. ,["ò", "o"],["ó", "o"],["ô", "o"],["õ", "o"],["ö", "o"],["ö", "o"],["œ", "oe"]  
 -. ,["Ù", "U"],["Ú", "U"],["Û", "U"],["Ü", "U"],["Ü", "U"]  
 -. ,["ù", "u"],["ú", "u"],["û", "u"],["ü", "u"],["ü", "u"]  
 -. ,["Ý", "Y"],["Ÿ", "Y"],["Ÿ", "Y"]  
 -. ,["ý", "y"],["ÿ", "y"],["ÿ", "y"]  
 -. ,["Þ", "P"],["þ", "p"],["ß", "s"]["¡", "!"]["¿", "?"]  
  -> 위 용어를 모두 분석한 결과, 부분적으로만 대문자로 시작되는 글자는 없었으며,    
  -> 제품의 발음을 위해 France 등 일부 유럽국가에서만 특수문자로 기재가 되어 있음을 확인하였다.   
  -> 이에, 카테고리티를 줄이기 위해서 이를 통일하는 작업이 필요하다가 생각된다. 

10. 출처   
 -. 와인등급 :https://www.wine21.com/11_news/news_view.html?Idx=17111  
 -. 와인등급 : https://m.blog.naver.com/interkj87/220699483883  
 -. 와인등급 : https://www.decantalo.com/en/blog/winemaking/classifying-wines-by-their-age-crianza-reserva-gran-reserva/  
 -. 유럽(문자) 표기법:http://help.hancom.com/hoffice/multi/ko_kr/hwp/insert/europe.htm
※ 모든 title에 해당 정보가 기재되어 있음

유럽(외국문자) 표기법
http://help.hancom.com/hoffice/multi/ko_kr/hwp/insert/europe.htm

# 2. EDA


   -. 
   -.   
  * Feature Engineering 
   -. description 분석   
   -. country기준으로 designation Features수 줄이기   

In [8]:
# 혼용되는 용어 통일을 위한 object 추출 
df.select_dtypes(exclude="number").columns

Index(['country', 'description', 'designation', 'province', 'region_1',
       'region_2', 'taster_name', 'taster_twitter_handle', 'title', 'variety',
       'winery'],
      dtype='object')

In [9]:
# 혼용되는 용어 통일을 위한 object 리스트화
object_columns = ['country', 'description', 'designation', 'province', 'region_1',
       'region_2', 'taster_name', 'taster_twitter_handle', 'title', 'variety',
       'winery']

In [10]:
df[object_columns].dtypes

country                  object
description              object
designation              object
province                 object
region_1                 object
region_2                 object
taster_name              object
taster_twitter_handle    object
title                    object
variety                  object
winery                   object
dtype: object

In [11]:
df[~df.isnull()]["variety"]

0            White Blend
1         Portuguese Red
2             Pinot Gris
3               Riesling
4             Pinot Noir
               ...      
129966          Riesling
129967        Pinot Noir
129968    Gewürztraminer
129969        Pinot Gris
129970    Gewürztraminer
Name: variety, Length: 129971, dtype: object

In [12]:
# 위에서도 언급한 바와 같이 같은 용어인데 나라별로 철자를 다르게 사용한다. 
# 이에, 용어를 통일하고자 한다. 

for column in df[~df.isnull()][object_columns]:
    df[column] = df[column].str.replace("reserva","reserve")
    df[column] = df[column].str.replace("riserva","reserve")

In [13]:
# "designation","description" -> 변경된 값이 모두 정상적으로 변경되어 출력되지 않음
df[["designation","description"]].query('description.str.contains("reserva")',engine='python').head(10)

Empty DataFrame
Columns: [designation, description]
Index: []

## 2-1) 특수문자 -> 영어로 변환하기

In [14]:
# 문자 영어로 통일 
# 위에서도 언급한 바와 같이 해당 데이터에는 발음의 구분 목적으로 
# 특수문자가 기재되어 있어 해당 데이터를 영어로 변환해주고자 함 

lists = [["À","Á","Â","Ã","Ä","Ä","Å"]
        ,["à","á","â","ã","ä","ä","å"]
        ,["È","É","Ê","Ë","Ë"]
        ,["è", "é", "ê", "ë", "ë"]     
        ,["ç", "c"]
        ,["ì","í","î","ï","ï"]
        ,["ò","ó","ô","õ","ö","ö","œ"]
        ,["ñ"]
        ,["ù","ú","û","ü","ü"]
        ,["ý","ÿ","ÿ"]]

for list in lists[0][1] :
    df = df.replace(list, "a", regex=True)

for list in lists[2][3] :
    df = df.replace(list, "e", regex=True)
    
for list in lists[4] :
    df = df.replace(list, "c", regex=True)
    
for list in lists[5] :
    df = df.replace(list, "i", regex=True)
    
for list in lists[6] :
    df = df.replace(list, "o", regex=True)
    
for list in lists[7] :
     df = df.replace(list, "n", regex=True)
    
for list in lists[8] :
    df = df.replace(list, "u", regex=True)
    
for list in lists[9] :
    df = df.replace(list, "y", regex=True)

In [15]:
# 모든 대문자 -> 소문자
for column in object_columns:
    df[column] = df[column].str.lower()

In [16]:
# 모두 제거 되고, 대문자로 소문자로 변경된 것을 확인 
df.tail(5)

country                                        description  \
129966  germany  notes of honeysuckle and cantaloupe sweeten th...   
129967       us  citation is given as much as a decade of bottl...   
129968   france  well-drained gravel soil gives this wine its c...   
129969   france  a dry style of pinot gris, this is crisp with ...   
129970   france  big, rich and off-dry, this is powered by inte...   

                                   designation  points  price province  \
129966  brauneberger juffer-sonnenuhr spätlese      90   28.0    mosel   
129967                                     NaN      90   75.0   oregon   
129968                                   kritt      90   30.0   alsace   
129969                                     NaN      90   32.0   alsace   
129970           lieu-dit harth cuvée caroline      90   21.0   alsace   

       region_1      region_2         taster_name taster_twitter_handle  \
129966      NaN           NaN  anna lee c. iijima                   NaN   
129967   oregon  oregon other        paul gregutt           @paulgwine    
129968   alsace           NaN          roger voss            @vossroger   
129969   alsace           NaN          roger voss            @vossroger   
129970   alsace           NaN          roger voss            @vossroger   

                                                    title         variety  \
129966  dr. h. thanisch (erben muller-burggraef) 2013 ...        riesling   
129967                  citation 2004 pinot noir (oregon)      pinot noir   
129968  domaine gresser 2013 kritt gewurztraminer (als...  gewurztraminer   
129969      domaine marcel deiss 2012 pinot gris (alsace)      pinot gris   
129970  domaine schoffit 2012 lieu-dit harth cuvée car...  gewurztraminer   

                                          winery  
129966  dr. h. thanisch (erben muller-burggraef)  
129967                                  citation  
129968                           domaine gresser  
129969                      domaine marcel deiss  
129970                          domaine schoffit

## 2-2) title를 통해 region 결측치 해결하기 

In [17]:
# "region_1","title" 관찰 
df[["region_1","title"]].query('title.str.contains("joven")',engine='python').head(10)

region_1                                              title
1827      ribeira sacra  adegas moure 2013 tradicion joven red (ribeira...
2138              rioja  bodegas riojanas 2015 canchales vino joven  (r...
4570              rioja           covila 2015 adivino tinto joven  (rioja)
7705   ribera del duero  pago de los capellanes 2006 joven roble  (ribe...
9360   ribera del duero              arroyo 2006 joven  (ribera del duero)
9421          la mancha  vinicola de castilla 2011 finca del castillo j...
16034              toro                    muruve 2009 joven tinto  (toro)
16295        terra alta  finca hispana 2013 joven garnacha blanca (terr...
21151       paso robles    croma vera 2015 joven tempranillo (paso robles)
25668  ribera del duero      pradorey 2011 tinto joven  (ribera del duero)

위와 같이 데이터를 살펴본 결과, title의 괄호 부분이 region_1를 나타냄을 알 수 있었다.
이로 region_1의 결측치 21,247개를 처리할 수 있다.

In [18]:
# 먼저 (로 split하고 1번째 인덱스를 region으로 저장 
df['region'] = df.title.str.split('(').str[1]

# 먼저 )로 split하고 1번째 인덱스를 region으로 저장 
df['region'] = df.region.str.split(')').str[0]
df['region']

0                           etna
1                          douro
2              willamette valley
3            lake michigan shore
4              willamette valley
                   ...          
129966    erben muller-burggraef
129967                    oregon
129968                    alsace
129969                    alsace
129970                    alsace
Name: region, Length: 129971, dtype: object

In [19]:
# 데어터 컬럼 생성여부 확인
df.head(1)

country                                        description   designation  \
0   italy  aromas include tropical fruit, broom, brimston...  vulkà bianco   

   points  price           province region_1 region_2    taster_name  \
0      87    NaN  sicily & sardinia     etna      NaN  kerin o’keefe   

  taster_twitter_handle                              title      variety  \
0          @kerinokeefe  nicosia 2013 vulkà bianco  (etna)  white blend   

    winery region  
0  nicosia   etna

In [20]:
df["title"].isnull().sum()

0

In [21]:
print(df['region_1'].isnull().sum())
print(df['region'].isnull().sum())

21247
10522


region_1보다는 결측치를 만개주렸다. title에는 결측치가 없으나, title내에 ()로 지역명이 기재되어 있지 않은 데이터가 1만개 정도 있는 것 같다 

## 2-3) title에서 연도 분리하기 

In [22]:
# "region_1","title" 관찰 
df[["title"]].query('title.str.contains("joven")',engine='python').head(10)

title
1827   adegas moure 2013 tradicion joven red (ribeira...
2138   bodegas riojanas 2015 canchales vino joven  (r...
4570            covila 2015 adivino tinto joven  (rioja)
7705   pago de los capellanes 2006 joven roble  (ribe...
9360               arroyo 2006 joven  (ribera del duero)
9421   vinicola de castilla 2011 finca del castillo j...
16034                    muruve 2009 joven tinto  (toro)
16295  finca hispana 2013 joven garnacha blanca (terr...
21151    croma vera 2015 joven tempranillo (paso robles)
25668      pradorey 2011 tinto joven  (ribera del duero)

In [58]:
df["title"][0]

'nicosia 2013 vulkà bianco  (etna)'

In [59]:
df["title"][1]

'quinta dos avidagos 2011 avidagos red (douro)'

title에는 와인 규정상 생성된 연도를 기재하게 되어 있다.  
이 특성을 이용하여 와인의 생성연도를 구해보고자 한다. 

In [85]:
for i in range(10):
    print(df["title"][i])

nicosia 2013 vulkà bianco  (etna)
quinta dos avidagos 2011 avidagos red (douro)
rainstorm 2013 pinot gris (willamette valley)
st. julian 2013 reserve late harvest riesling (lake michigan shore)
sweet cheeks 2012 vintner's reserve wild child block pinot noir (willamette valley)
tandem 2011 ars in vitro tempranillo-merlot (navarra)
terre di giurfo 2013 belsito frappato (vittoria)
trimbach 2012 gewurztraminer (alsace)
heinz eifel 2013 shine gewurztraminer (rheinhessen)
jean-baptiste adam 2012 les natures pinot gris (alsace)


In [81]:
year_wine = []

In [55]:
df["title"].isnull().sum()

0

In [88]:
# year feature 생성 
title_lists = df["title"].tolist()
for list in title_lists :
    test = re.findall("\d+", list)
    n = 0 
    for i in test:
        n = 0 
        if int(i) > 1990 :
            df["year"][n] = i
        else :
            df["year"][n] = 0
        n = n+1

<ipython-input-88-c270cbedbc91>:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["year"][n] = i
<ipython-input-88-c270cbedbc91>:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["year"][n] = 0


In [89]:
len(year_wine)

129494

In [90]:
df.shape

(129970, 19)

In [91]:
year_wine

[2013,
 2013,
 2010,
 2013,
 2007,
 2015,
 2010,
 2013,
 2012,
 2011,
 2010,
 2013,
 2010,
 2015,
 2011,
 2013,
 2006,
 2007,
 2014,
 2014,
 2006,
 2012,
 2013,
 2011,
 2010,
 2007,
 2011,
 2006,
 2013,
 2010,
 2010,
 2010,
 2011,
 2012,
 2010,
 2008,
 2000,
 2010,
 2013,
 2011,
 2008,
 2015,
 2007,
 2010,
 2007,
 0,
 2008,
 2007,
 2010,
 2011,
 2007,
 2010,
 2011,
 2013,
 2010,
 2009,
 2016,
 0,
 2016,
 0,
 2015,
 2016,
 2015,
 2013,
 2012,
 2012,
 2016,
 2010,
 2008,
 2013,
 2015,
 2016,
 2015,
 2009,
 2011,
 2014,
 2012,
 2011,
 2006,
 2016,
 2015,
 2013,
 0,
 2015,
 2008,
 2010,
 2012,
 2006,
 2013,
 2003,
 2015,
 2015,
 2015,
 0,
 2009,
 2011,
 2011,
 2009,
 2012,
 2007,
 2008,
 2010,
 2011,
 2011,
 2015,
 2012,
 2006,
 2012,
 2011,
 2010,
 2006,
 2006,
 2006,
 2008,
 2014,
 2014,
 2015,
 2013,
 2013,
 2015,
 2013,
 0,
 2008,
 2015,
 2007,
 2014,
 2014,
 2015,
 2012,
 2016,
 2007,
 0,
 0,
 2013,
 0,
 1999,
 2007,
 2008,
 2011,
 2011,
 2012,
 2006,
 2005,
 2011,
 2013,
 2000,
 2008

In [92]:
df.head(50)

variety                winery country  \
0       white blend               nicosia   italy   
12551   white blend               nicosia   italy   
52317   white blend               nicosia   italy   
12546   white blend   masseria setteporte   italy   
22      white blend    baglio di pianetto   italy   
76833   white blend    baglio di pianetto   italy   
123206  white blend    baglio di pianetto   italy   
26      white blend              stemmari   italy   
22394   white blend              stemmari   italy   
32      white blend    duca di salaparuta   italy   
10113   white blend    duca di salaparuta   italy   
13958   white blend    duca di salaparuta   italy   
109873  white blend    duca di salaparuta   italy   
112309  white blend    duca di salaparuta   italy   
126504  white blend    duca di salaparuta   italy   
13780   white blend               estampa   chile   
28315   white blend               estampa   chile   
58292   white blend               estampa   chile   
102675  white blend               estampa   chile   
109324  white blend               estampa   chile   
116342  white blend               estampa   chile   
117239  white blend               estampa   chile   
46658   white blend   feudi del pisciotto   italy   
96437   white blend   feudi del pisciotto   italy   
116943  white blend   feudi del pisciotto   italy   
38570   white blend  feudo di santa tresa   italy   
61228   white blend  feudo di santa tresa   italy   
77441   white blend  feudo di santa tresa   italy   
109613  white blend  feudo di santa tresa   italy   
1959    white blend                tarara      us   
14238   white blend                tarara      us   
46953   white blend                tarara      us   
58148   white blend                tarara      us   
103008  white blend                tarara      us   
103803  white blend                tarara      us   
106481  white blend                tarara      us   
12244   white blend      tasca d'almerita   italy   
13504   white blend      tasca d'almerita   italy   
17187   white blend      tasca d'almerita   italy   
20676   white blend      tasca d'almerita   italy   
30771   white blend      tasca d'almerita   italy   
45368   white blend      tasca d'almerita   italy   
50506   white blend      tasca d'almerita   italy   
63880   white blend      tasca d'almerita   italy   
67717   white blend      tasca d'almerita   italy   
98770   white blend      tasca d'almerita   italy   
102574  white blend      tasca d'almerita   italy   
50668   white blend            casa silva   chile   
2425    white blend                 corvo   italy   
6416    white blend                 corvo   italy   

                                              description  \
0       [aromas, include, tropical, fruit,, broom,, br...   
12551   [a, delicate, floral, fragrance, of, white, fl...   
52317   [a, blend, of, carricante, and, catarratto, (t...   
12546   [aromas, of, spanish, broom,, acacia, and, whi...   
22      [delicate, aromas, recall, white, flower, and,...   
76833   [a, blend, of, 70%, inzolia, and, 30%, viognie...   
123206  [this, luminous, blend, of, insolia, and, viog...   
26      [pretty, aromas, of, yellow, flower, and, ston...   
22394   [a, blend, of, 80%, grillo, and, 20%, viognier...   
32      [part, of, the, extended, calanica, series,, t...   
10113   [this, value, wine, opens, with, a, fresh,, ci...   
13958   [a, blend, of, insolia, and, chardonnay,, this...   
109873  [the, blend, is, 60-40, grillo, and, viognier,...   
112309  [this, offers, subdued, aromas, suggesting, or...   
126504  [this, blend, of, insolia, and, chardonnay, op...   
13780   [this, is, a, vegetal,, underripe, wine,, albe...   
28315   [the, bouquet, smells, like, spiked, fruit, co...   
58292   [starts, out, neutral, but, soon, it, develops...   
102675  [bland,, pithy, citrus, aromas, come, with, no...   
109324  [pickled,, briny, aromas, lead, to, a, palate,...   
116342  [the, bouquet, smel

In [65]:
df["year"].unique()

array([2008, 2014, 2012,    0], dtype=int64)

In [24]:
# year생성 여부 확인 
df["year"].unique()

array([2012], dtype=int64)

## 2-4) winery에 가중치 부여하기 

In [52]:
df["winery"][0][1]

'i'

In [53]:
df["winery"]

0                     nicosia
12551                 nicosia
52317                 nicosia
12546     masseria setteporte
22         baglio di pianetto
                 ...         
101010       terra de falanis
113067    suertes del marqués
105149            villa venti
119787      amiran vepkhvadze
127959                  emery
Name: winery, Length: 129970, dtype: object

In [25]:
df["winery"].describe()

count                 129971
unique                 16728
top       wines & winemakers
freq                     222
Name: winery, dtype: object

In [26]:
df["winery"].isnull().sum()

0

해당 데이터를 살펴보니 unique는 높으나 결측치가 없는 것을 확인된다.  
이에 해당 데이터를 카테고리별로 카운트하여 총 카운트된 갯수로 나누어준 점수로 구성된  
새로운 feature를 만들어주고자 한다.

위와 같이 처리를 하였을 때 좋은 점은 개별점수가 하나의 개별 unique가 되면서 회귀분석을 진행할 수 있다는 점이다. 

In [27]:
# winery value_count 및 score 생성 
winery_count = pd.DataFrame(df["winery"].value_counts())
winery_count["winery"] = winery_count["winery"]/len(winery_count["winery"])
winery_count.head(2)

winery
wines & winemakers  0.013271
testarossa          0.013032

In [28]:
# data wrangling을 통한 카테고리데이터 점수화 및 merge
df = pd.merge(df, winery_count, left_on="winery", right_index=True)
df.head(1)

# 카피가 되면서 merge가 되어 추후 필요없는 feature는 일괄적으로 삭제 예정

winery country                                        description  \
0  nicosia   italy  aromas include tropical fruit, broom, brimston...   

    designation  points  price           province region_1 region_2  \
0  vulkà bianco      87    NaN  sicily & sardinia     etna      NaN   

     taster_name taster_twitter_handle                              title  \
0  kerin o’keefe          @kerinokeefe  nicosia 2013 vulkà bianco  (etna)   

       variety winery_x region  year  winery_y  
0  white blend  nicosia   etna  2012  0.000658

In [29]:
df["winery_y"].isnull().sum()

0

## 2-5) winery에 가중치 부여하기 

In [30]:
df['variety'].value_counts().head(5)

pinot noir                  13272
chardonnay                  11753
cabernet sauvignon           9472
red blend                    8946
bordeaux-style red blend     6915
Name: variety, dtype: int64

카테고리티가 너무 높기 때문에 이를 이대로 인코더를 통해 학습을 진행할 경우 문제가 feature가 너무나 많아지게 된다. 이에, 해결하기 위해 아래와 같이 2가지를 고민하였다.

구간별로 그룹을 지어 카테고리별로 분리를 할 것인지
count / len로 각 카테고리별 점수표를 만들어서 가중치를 줄것인지
1번을 선택할 경우, 그룹을 나누는 기준을 선정하는 것이 매우 모호하다
수량으로 기준을 나누자니 아래 variety_point에서 관찰되듯 variety_count가 높다고 무조건 max score가 높은 것도 아니며, 평균이 높은 것도 아니다. 이에, 2번 방법으로 각 카테코리별 점수표를 만들어 가중치를 주고자 한다.

위와 같이 처리를 하였을 때 좋은 점은 개별점수가 하나의 개별 unique가 되면서 회귀분석을 진행할 수 있다는 점이다. 

In [31]:
variety_point = df.groupby(['variety']).points.agg([len, min, max, sum])
variety_point["mean"] = variety_point["sum"]/variety_point["len"]
variety_point.sort_values(by="len", ascending=False).head(5)

len  min  max      sum       mean
variety                                                      
pinot noir                13272   80   99  1186669  89.411468
chardonnay                11753   80  100  1038261  88.340083
cabernet sauvignon         9472   80  100   839291  88.607580
red blend                  8946   80   99   790650  88.380282
bordeaux-style red blend   6915   80  100   616171  89.106435

In [32]:
variety_count = pd.DataFrame(df["variety"].value_counts())
variety_count["variety"] = variety_count["variety"]/len(variety_count["variety"])
variety_count.shape # 누락된 값있는지 확인 

(707, 1)

In [33]:
# 생성된 variety_count vs variety unique 비교 
df["variety"].describe()

count         129970
unique           707
top       pinot noir
freq           13272
Name: variety, dtype: object

In [34]:
# variety를 카운트하여 해당 카테고리를 인트로 변환하여 주었다. 
# 추후 scaler를 통하여 해당 값들은 조정해줄 예정이다. 
df = pd.merge(df, variety_count, left_on="variety", right_index=True)
df.head(2)

# 카피가 되면서 merge가 되어 추후 필요없는 feature는 일괄적으로 삭제 예정

variety   winery country  \
0      white blend  nicosia   italy   
12551  white blend  nicosia   italy   

                                             description          designation  \
0      aromas include tropical fruit, broom, brimston...         vulkà bianco   
12551  a delicate floral fragrance of white flower, p...  fondo filara bianco   

       points  price           province region_1 region_2    taster_name  \
0          87    NaN  sicily & sardinia     etna      NaN  kerin o’keefe   
12551      88    NaN  sicily & sardinia     etna      NaN  kerin o’keefe   

      taster_twitter_handle                                     title  \
0              @kerinokeefe         nicosia 2013 vulkà bianco  (etna)   
12551          @kerinokeefe  nicosia 2013 fondo filara bianco  (etna)   

         variety_x winery_x region  year  winery_y  variety_y  
0      white blend  nicosia   etna  2012  0.000658   3.338048  
12551  white blend  nicosia   etna  2012  0.000658   3.338048

## 2-6) description 확인하기

In [35]:
# description을 구성하는 단어 살펴보기 
df["description"] = df["description"].str.split(" ")
lists = df["description"].tolist()

all_list = []
for list in lists:
    for item in list:
        all_list.append(item)

In [36]:
# 단어가 각각 총 몇번 언급되었는지 확인 
counts = {}
for list in all_list :
    counts[list] = counts.get(list, 0) + 1
    
{k: v for k, v in sorted(counts.items(), key=lambda item: item[1], reverse=True)}

{'and': 347196,
 'the': 220799,
 'a': 179316,
 'of': 172882,
 'with': 120170,
 'this': 113583,
 'is': 96263,
 'in': 61814,
 'wine': 56149,
 'to': 55421,
 'it': 54326,
 'on': 45193,
 'flavors': 41815,
 'that': 36361,
 'aromas': 35267,
 'from': 32481,
 'but': 31010,
 'fruit': 30011,
 "it's": 29886,
 'drink': 26845,
 'palate': 26193,
 'are': 25784,
 'black': 25235,
 'has': 24691,
 'for': 21592,
 'ripe': 21520,
 'by': 20632,
 'acidity': 19611,
 'finish.': 19230,
 'tannins': 18283,
 'as': 18175,
 'an': 18149,
 'its': 17951,
 'red': 17731,
 'cherry': 16892,
 'notes': 15048,
 'flavors.': 13979,
 'fresh': 13021,
 'finish': 12716,
 'offers': 12648,
 'rich': 12496,
 'spice': 11904,
 'white': 11677,
 'while': 11676,
 'shows': 11449,
 'berry': 11449,
 'blend': 11346,
 'acidity.': 11212,
 'wine,': 10927,
 'wine.': 10779,
 'through': 10768,
 'at': 10509,
 'very': 10209,
 'nose': 9881,
 'dark': 9874,
 'plum': 9808,
 'dry': 9662,
 'tannins.': 9628,
 'crisp': 9624,
 'more': 9531,
 'oak': 9527,
 'cherry

위와 같이 description을 분석한 결과, 검색되는 데이터량이 너무 많고 이들 중 wine을 대표할만한 키워드를 찾는다하여도 feature수가 급증하며, 문장으로써 의미를 갖는 것에 대한 가중치를 계산하는 것이 불가능하므로 해당 feature는 분석에서 제외토록 하겠습니다. 

## 2-7) designation 확인하기 - feature groupby 시도

In [37]:
# 최초 unique(37979)에서 특수문자 & 용어통일작업으로  unique(37609)로 감소 ( 감소량 : 370)
df["designation"].describe()

count       92506
unique      37609
top       reserve
freq         2009
Name: designation, dtype: object

위에서 언급한 등급으로 해당 feature를 그룹화해보고자 한다. 

In [38]:
# # 문자를 replace하는데 조금 더 효율적으로 변환하기 위해 공백을 _로 변경
# df["designation"] = df["designation"].str.replace(" ", "_")

# # 결측치가 있는 경우, replace 오류가 발생됨 
# # 이에, ~ 를 통해 결측치 있는 부분은 제외하고 designation을 선택하여 
# # 등급별로 변환하는 작업을 진행함 
# for grade in df.loc[~df["designation"].isnull()]["designation"]:
#     if "joven" in grade:
#         df["designation"] = df["designation"].str.replace(grade, "joven")
#     if "crianza" in grade:
#         df["designation"] = df["designation"].str.replace(grade, "crianza")
#     if "gran_reserve" in grade:
#         df["designation"] = df["designation"].str.replace(grade, "gran_reservA")
#     if "reserve" in grade:
#         df["designation"] = df["designation"].str.replace(grade, "reserve")

위와 같이 처리를 하는 것은 매우 시간이 오래걸리고, unique가 1만개정도 밖에 감소를 하지 않기에 해당방법은 사용하지 않도록 하겠습니다. 담당자가 생각한 것과는 다르게 designation를 모두 그룹화 되지 못하였습니다. 
이에, 각 카테고리별 가중치를 부여하여 해당 문제를 해결하고 합니다. ( 시간이 오래걸려 해당 CODE는 주석처리함 )

In [39]:
df.describe(include='number')

points          price      year       winery_y      variety_y
count  129970.000000  120974.000000  129970.0  129970.000000  129970.000000
mean       88.447142      35.363541    2012.0       0.001719       7.581686
std         3.039742      41.022353       0.0       0.001947       6.424914
min        80.000000       4.000000    2012.0       0.000060       0.001414
25%        86.000000      17.000000    2012.0       0.000418       1.902405
50%        88.000000      25.000000    2012.0       0.001076       5.041018
75%        91.000000      42.000000    2012.0       0.002331      13.397454
max       100.000000    3300.000000    2012.0       0.013271      18.772277

In [40]:
df["year"]

0         2012
12551     2012
52317     2012
12546     2012
22        2012
          ... 
101010    2012
113067    2012
105149    2012
119787    2012
127959    2012
Name: year, Length: 129970, dtype: int64